plugins/gradle/tooling-extension-impl/src/org/jetbrains/plugins/gradle/tooling/tasks/DependenciesReport.groovy

```diff
@@ -54,191 +53,146 @@ class DependenciesReport extends DefaultTask {
+            }
           }
         }
       }
-    }
 
-    Map<Object, DependencyNode> added = [:]
-    root.getChildren().each { RenderableDependency child ->
-      node.dependencies.add(toNode(child, configurationName, added, idGenerator, projectNameFunction))
-    }
-    return node
-  }
+      Map<Object, DependencyNode> added = [:]
 
-  static private DependencyNode toNode(RenderableDependency dependency,
-                                       String configurationName,
-                                       Map<Object, DependencyNode> added,
-                                       IdGenerator idGenerator,
-                                       ProjectNameFunction projectNameFunction) {
-    long id = idGenerator.getId(dependency, configurationName)
-    DependencyNode alreadySeenNode = added.get(id)
-    if (alreadySeenNode != null) {
-      return new ReferenceNode(id)
-    }
-
-    AbstractDependencyNode node
-    if (dependency.id instanceof ProjectComponentIdentifier) {
-      ProjectComponentIdentifier projectId = dependency.id as ProjectComponentIdentifier
-      node = new ProjectDependencyNodeImpl(id, projectNameFunction.fun(projectId))
-    }
-    else if (dependency.id instanceof ModuleComponentIdentifier) {
-      ModuleComponentIdentifier moduleId = dependency.id as ModuleComponentIdentifier
-      node = new ArtifactDependencyNodeImpl(id, moduleId.group, moduleId.module, moduleId.version)
-    }
-    else {
-      node = new UnknownDependencyNode(id, dependency.name)
-    }
-    node.setResolutionState(dependency.resolutionState.name())
-    if (dependency instanceof CustomRenderableDependencyResult) {
-      ComponentSelectionReason selectionReason = dependency.selectionReason
-        if (!selectionReason.descriptions.isEmpty()) {
-          node.selectionReason = selectionReason.descriptions.last().description
-        }
-    }
-    added.put(id, node)
-    Iterator<? extends RenderableDependency> iterator = dependency.getChildren().iterator()
-    while (iterator.hasNext()) {
-      RenderableDependency child = iterator.next()
-      node.dependencies.add(toNode(child, configurationName, added, idGenerator, projectNameFunction))
-    }
-    return node
-  }
-
-  static class ProjectNameFunction {
-    def is45OrNewer = GradleVersion.current() >= GradleVersion.version("4.5")
-
-    String fun(ProjectComponentIdentifier identifier) {
-      return is45OrNewer ? identifier.projectName : identifier.projectPath
-    }
-  }
-
-  private static class IdGenerator {
-    private Map<String, Long> idMap = new HashMap<>()
-    private long value
-
-    private long getId(String prefix, String configurationName) {
-      def key = prefix + '_' + configurationName
-      def id = idMap.get(key)
-      if (id == null) {
-        idMap[key] = ++value
-        id = value
+      def iterator = root.dependencies.iterator()
+      while (iterator.hasNext()) {
```

Comment:
FYI this weird iterator scheme is the only way to get fast iteration in Groovy. It would probably be best to rewrite this class in Java at some point.